algorithms/dp/bitmasking_tsp.py

```diff
@@ -1,15 +1,187 @@
+                if not visited[curr_row][curr_col]:
+                    visited[curr_row][curr_col] = True
+                    dist[curr_row][curr_col][index] = dist[pos[0]][pos[1]][index] + 1
+                    queue.append([curr_row, curr_col])
+
+"""
+Compute all the distances from all houses to all nodes
+
+:param nodes: The nodes
+:param houses: The houses location
+:param nbRow: The number of rows
+:param nbColumn: The number of columns
+:returns: All the distances
+"""
+def getAllDist(nodes: list, houses: list, nbRow: int, nbColumn: int):
+
+    dist = [[[math.inf for house in range(len(houses))] for column in range(nbColumn)] for row in range(nbRow)]
+
+    for index in range(len(houses)):
+        getDist(nodes, houses, nbRow, nbColumn, dist, index)
+
+    return dist
+
+"""
+Initialize the variables
+
+:param nodes: The nodes given
+:param nbRow: The number of rows
+:param nbColumn: The number of columns
+"""
+def initialization(nodes: list, nbRow: int, nbColumn: int):
+
+    # check if the arguments are valid
+    check_argument(nodes, nbRow, nbColumn)
+
+    # add starting point as "house"
+    houses = [[0, 0]]
+
+    # get all houses
+    for i in range(nbRow):
+        for j in range(nbColumn):
+            if nodes[i][j] == '*':
+                houses.append([i, j])
+    
+    # check the number of houses
+    if len(houses) > MAX_HOUSE:
+        raise ValueError(f'The maximum number of houses is {MAX_HOUSE}, but you gave {len(houses)} houses !')
+    
+    # initialize the DP matrix
+    DP = [[-1 for mask in range(2 ** len(houses))] for i in range(len(houses))]
+    
+    # compute all distances
+    dist = getAllDist(nodes, houses, nbRow, nbColumn)
+
+    return houses, DP, dist
+
+"""
+Find the shortest path
+
+:param houses: The houses location
+:param DP: The DP matrix
+:param dist: All distances
+:param index: The house index
+:param mask: The current mask
+:returns: The shortest path
+"""
+def find_shortest_path(houses: list, DP: list, dist: list, index: int, mask: int):
+
+    # if the mask is full, then we found a path
+    if mask == (1 << len(houses)) - 1:
+        return dist[0][0][index]
+    
+    # if this subproblem has already been computed, just return the solution
+    if DP[index][mask] != -1:
+        return DP[index][mask]
+    
+    minDist = math.inf
+
+    for index_ in range(len(houses)):
+        if not (mask & (1 << index_)):
+            minDist = min(
+                minDist,
+                find_shortest_path(houses, DP, dist, index_, mask | (1 << index_)) + dist[houses[index_][0]][houses[index_][1]][index]
+            )
+    
+    # store the result in the DP matrix
+    DP[index][mask] = minDist
+
+    return DP[index][mask]
+
 """
 Finds the length of the shortest Euler circuit using bitmasking and dynamic programming (DP).
 :param nodes: Contains integer tuples of the node coordinates.
 :returns: the length of the shortest Euler circuit.
```

Comment:
Could you update the comment by including the other two params and making sure the description of _nodes_ is correct?

Reply:
Sure, I forgot to update it, good catch !